File: robot_server/client/cpp/smolvla_client.cpp
#include "smolvla_client.h"

#include "socket.h"

#include <cstring>
#include <utility>

namespace proto = smolvla::protocol;
namespace sockets = robot_server::sockets;

namespace robot_server {
namespace client {
namespace {

bool send_message(
    sockets::socket_handle fd,
    uint16_t op,
    uint32_t request_id,
    const std::vector<uint8_t> & payload,
    std::string & error) {
    proto::header h;
    h.op = op;
    h.request_id = request_id;
    h.payload_len = payload.size();

    std::vector<uint8_t> header;
    proto::encode_header(h, header);
    if (!sockets::send_all(fd, header.data(), header.size(), error)) {
        return false;
    }
    if (!payload.empty() && !sockets::send_all(fd, payload.data(), payload.size(), error)) {
        return false;
    }
    return true;
}

bool recv_message(
    sockets::socket_handle fd,
    proto::header & h,
    std::vector<uint8_t> & payload,
    std::string & error) {
    std::vector<uint8_t> header(proto::k_header_size);
    if (!sockets::recv_exact(fd, header.data(), header.size(), error)) {
        return false;
    }
    if (!proto::decode_header(header.data(), header.size(), h, error)) {
        return false;
    }
    if (!proto::validate_header(h, proto::k_default_max_payload, error)) {
        return false;
    }

    payload.assign((size_t) h.payload_len, 0);
    if (!payload.empty() && !sockets::recv_exact(fd, payload.data(), payload.size(), error)) {
        return false;
    }
    return true;
}

bool make_predict_request(const SmolVLAObservation & obs, proto::predict_request & req, std::string & error) {
    if (!obs.rgb_hwc_u8 || obs.width == 0 || obs.height == 0) {
        error = "observation image must be non-empty RAW_RGB_U8";
        return false;
    }

    const uint32_t stride = obs.stride_bytes == 0 ? obs.width * 3 : obs.stride_bytes;
    if (stride < obs.width * 3) {
        error = "observation stride_bytes is smaller than width * 3";
        return false;
    }

    req.image_format = proto::image_raw_rgb_u8;
    req.width = obs.width;
    req.height = obs.height;
    req.channels = 3;
    req.stride_bytes = stride;
    req.task = obs.prompt;
    if (obs.state && obs.state_dim > 0) {
        req.state.assign(obs.state, obs.state + obs.state_dim);
    } else {
        req.state.clear();
    }

    const size_t image_size = (size_t) stride * (size_t) obs.height;
    req.image.resize(image_size);
    std::memcpy(req.image.data(), obs.rgb_hwc_u8, image_size);
    return true;
}

} // namespace

const float * SmolVLAResponse::action_row(uint32_t index) const {
    if (action_dim == 0 || index >= chunk_size) {
        return nullptr;
    }
    const size_t offset = (size_t) index * (size_t) action_dim;
    if (offset + action_dim > actions_flat.size()) {
        return nullptr;
    }
    return actions_flat.data() + offset;
}

SmolVLAClient::SmolVLAClient(std::string host, uint16_t port) : host_(std::move(host)), port_(port) {}

bool SmolVLAClient::health(std::string & text, std::string & error) {
    std::vector<uint8_t> payload;
    if (!call(proto::op_health, {}, payload, error)) {
        return false;
    }
    return proto::decode_text_payload(payload, text, error);
}

bool SmolVLAClient::reset(std::string & text, std::string & error) {
    std::vector<uint8_t> payload;
    if (!call(proto::op_reset, {}, payload, error)) {
        return false;
    }
    return proto::decode_text_payload(payload, text, error);
}

bool SmolVLAClient::shutdown(std::string & text, std::string & error) {
    std::vector<uint8_t> payload;
    if (!call(proto::op_shutdown, {}, payload, error)) {
        return false;
    }
    return proto::decode_text_payload(payload, text, error);
}

bool SmolVLAClient::predict(const SmolVLAObservation & obs, SmolVLAResponse & response, std::string & error) {
    proto::predict_request req;
    if (!make_predict_request(obs, req, error)) {
        return false;
    }

    std::vector<uint8_t> request_payload;
    if (!proto::encode_predict_request(req, request_payload, error)) {
        return false;
    }

    std::vector<uint8_t> response_payload;
    if (!call(proto::op_predict, request_payload, response_payload, error)) {
        return false;
    }

    proto::predict_response resp;
    if (!proto::decode_predict_response(response_payload, resp, error)) {
        return false;
    }

    response.chunk_size = resp.chunk_size;
    response.action_dim = resp.action_dim;
    response.actions_flat = std::move(resp.actions);
    response.timing = resp.timing;
    return true;
}

bool SmolVLAClient::call(
    uint16_t op,
    const std::vector<uint8_t> & request_payload,
    std::vector<uint8_t> & response_payload,
    std::string & error) {
    if (!sockets::startup(error)) {
        return false;
    }

    sockets::socket_handle fd = sockets::tcp_connect(host_.c_str(), port_, error);
    if (fd == sockets::invalid_socket) {
        sockets::cleanup();
        return false;
    }

    const uint32_t request_id = next_request_id_++;
    bool ok = send_message(fd, op, request_id, request_payload, error);
    proto::header response_header;
    if (ok) {
        ok = recv_message(fd, response_header, response_payload, error);
    }
    sockets::close(fd);
    sockets::cleanup();
    if (!ok) {
        return false;
    }

    if (response_header.request_id != request_id) {
        error = "response request_id mismatch";
        return false;
    }
    if (response_header.status != proto::status_ok) {
        std::string text;
        std::string decode_error;
        proto::decode_text_payload(response_payload, text, decode_error);
        error = text.empty() ? "server returned error" : text;
        return false;
    }
    return true;
}

} // namespace client
} // namespace robot_server

File: robot_server/client/python/smolvla_client.py
#!/usr/bin/env python3

from __future__ import annotations

import socket
import struct
from dataclasses import dataclass
from typing import Any, Iterable, Sequence


MAGIC = 0x414C5653
VERSION = 1
HEADER_SIZE = 32

OP_HEALTH = 1
OP_RESET = 2
OP_PREDICT = 3
OP_SHUTDOWN = 4

STATUS_OK = 0
IMAGE_RAW_RGB_U8 = 1

HEADER = struct.Struct("<IHHHHIIQI")
PREDICT_REQ_FIXED = struct.Struct("<IIIIIIIQ")
PREDICT_RESP_FIXED = struct.Struct("<III9d")


@dataclass
class SmolVLAResponse:
    chunk_size: int
    action_dim: int
    actions_flat: list[float]
    timings: dict[str, float]

    @property
    def actions(self) -> list[list[float]]:
        return [
            self.actions_flat[i : i + self.action_dim]
            for i in range(0, len(self.actions_flat), self.action_dim)
        ]


def _recv_all(sock: socket.socket, n: int) -> bytes:
    chunks: list[bytes] = []
    remaining = n
    while remaining:
        chunk = sock.recv(remaining)
        if not chunk:
            raise RuntimeError("server closed connection")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _send_message(sock: socket.socket, op: int, request_id: int, payload: bytes = b"") -> None:
    sock.sendall(HEADER.pack(MAGIC, VERSION, HEADER_SIZE, op, 0, request_id, STATUS_OK, len(payload), 0))
    if payload:
        sock.sendall(payload)


def _recv_message(sock: socket.socket) -> tuple[int, int, int, bytes]:
    raw_header = _recv_all(sock, HEADER_SIZE)
    magic, version, header_size, op, _flags, request_id, status, payload_len, _reserved = HEADER.unpack(raw_header)
    if magic != MAGIC:
        raise RuntimeError(f"bad magic: 0x{magic:08x}")
    if version != VERSION:
        raise RuntimeError(f"bad version: {version}")
    if header_size != HEADER_SIZE:
        raise RuntimeError(f"bad header size: {header_size}")
    payload = _recv_all(sock, payload_len) if payload_len else b""
    return op, request_id, status, payload


def encode_predict_request(
    rgb_hwc_u8: bytes,
    width: int,
    height: int,
    state: Sequence[float],
    prompt: str,
    stride_bytes: int | None = None,
) -> bytes:
    channels = 3
    stride = stride_bytes if stride_bytes is not None and stride_bytes > 0 else width * channels
    prompt_bytes = prompt.encode("utf-8")
    payload = bytearray()
    payload += PREDICT_REQ_FIXED.pack(
        IMAGE_RAW_RGB_U8,
        width,
        height,
        channels,
        stride,
        len(state),
        len(prompt_bytes),
        len(rgb_hwc_u8),
    )
    for value in state:
        payload += struct.pack("<f", float(value))
    payload += prompt_bytes
    payload += rgb_hwc_u8
    return bytes(payload)


def decode_predict_response(payload: bytes) -> SmolVLAResponse:
    fixed_size = PREDICT_RESP_FIXED.size
    chunk_size, action_dim, action_count, *timing_values = PREDICT_RESP_FIXED.unpack(payload[:fixed_size])
    if action_count != chunk_size * action_dim:
        raise RuntimeError("bad action shape in server response")
    actions_flat = list(struct.unpack(f"<{action_count}f", payload[fixed_size : fixed_size + action_count * 4]))
    timing_names = [
        "vision_ms",
        "state_proj_ms",
        "llm_ms",
        "kv_extract_ms",
        "phase2_ms",
        "model_total_ms",
        "server_recv_ms",
        "server_queue_ms",
        "server_predict_ms",
    ]
    return SmolVLAResponse(chunk_size, action_dim, actions_flat, dict(zip(timing_names, timing_values)))


def image_to_rgb_hwc_u8_bytes(image: Any) -> tuple[bytes, int, int, int]:
    if isinstance(image, dict):
        rgb = image.get("rgb_hwc_u8")
        if rgb is None:
            rgb = image.get("bytes")
        width = image.get("width")
        height = image.get("height")
        if rgb is None or width is None or height is None:
            raise ValueError("raw image dict must contain rgb_hwc_u8, width, and height")
        stride = image.get("stride_bytes", int(width) * 3)
        return bytes(rgb), int(width), int(height), int(stride)

    if isinstance(image, (bytes, bytearray, memoryview)):
        raise ValueError("raw bytes need width/height metadata; pass an observation dict instead")

    if hasattr(image, "detach"):
        image = image.detach().cpu().numpy()

    shape = tuple(image.shape)
    if len(shape) != 3:
        raise ValueError(f"expected 3D RGB image, got shape={shape}")

    if shape[0] == 3 and shape[2] != 3:
        image = image.transpose(1, 2, 0)
        shape = tuple(image.shape)
    if shape[2] != 3:
        raise ValueError(f"expected RGB image with 3 channels, got shape={shape}")

    if str(image.dtype) != "uint8":
        image = (image.clip(0, 1) * 255).astype("uint8")

    height, width, channels = image.shape
    return image.tobytes(order="C"), int(width), int(height), int(width * channels)


def state_to_list(state: Any) -> list[float]:
    if state is None:
        return []
    if hasattr(state, "detach"):
        state = state.detach().cpu().numpy()
    if hasattr(state, "reshape"):
        state = state.reshape(-1)
    return [float(x) for x in state]


class SmolVLAClient:
    def __init__(self, host: str = "127.0.0.1", port: int = 5555, timeout: float | None = None):
        self.host = host
        self.port = port
        self.timeout = timeout
        self._request_id = 0

    def _next_request_id(self) -> int:
        self._request_id += 1
        return self._request_id

    def _call(self, op: int, payload: bytes = b"") -> bytes:
        with socket.create_connection((self.host, self.port), timeout=self.timeout) as sock:
            request_id = self._next_request_id()
            _send_message(sock, op, request_id, payload)
            _op, response_id, status, response = _recv_message(sock)
        if response_id != request_id:
            raise RuntimeError(f"response id mismatch: {response_id} != {request_id}")
        if status != STATUS_OK:
            raise RuntimeError(response.decode("utf-8", errors="replace"))
        return response

    def health(self) -> str:
        return self._call(OP_HEALTH).decode("utf-8", errors="replace")

    def reset(self) -> str:
        return self._call(OP_RESET).decode("utf-8", errors="replace")

    def shutdown(self) -> str:
        return self._call(OP_SHUTDOWN).decode("utf-8", errors="replace")

    def predict_raw_rgb(
        self,
        rgb_hwc_u8: bytes,
        width: int,
        height: int,
        state: Iterable[float],
        prompt: str = "grab the block.",
        stride_bytes: int | None = None,
    ) -> SmolVLAResponse:
        payload = encode_predict_request(rgb_hwc_u8, width, height, list(state), prompt, stride_bytes)
        return decode_predict_response(self._call(OP_PREDICT, payload))

    def predict(self, observation: dict[str, Any], prompt: str | None = None) -> SmolVLAResponse:
        image = observation.get("image")
        if image is None:
            image = observation.get("observation.image")
        if image is None:
            raise ValueError("observation must contain 'image' or 'observation.image'")

        state = observation.get("state")
        if state is None:
            state = observation.get("observation.state")

        if prompt is None:
            prompt = observation.get("prompt")
        if prompt is None:
            prompt = observation.get("task", "grab the block.")

        rgb, width, height, stride = image_to_rgb_hwc_u8_bytes(image)
        return self.predict_raw_rgb(
            rgb,
            width=width,
            height=height,
            stride_bytes=stride,
            state=state_to_list(state),
            prompt=prompt,
        )
